### PR TITLE
 Add v3 RPC for sending cluster states to content nodes

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review
@toregge FYI

Supports sending states for multiple bucket spaces at once, with
possibilities for extending protocol later for sending distribution
config etc. Payload is LZ4-compressed when possible.

If a receiver node does not understand the v3 protocol, the protocol
version is transparently downgraded to legacy v2 RPC.

### CODE REVIEW

File: clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/ClusterStateBundleCodec.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core.rpc;
+
+import com.yahoo.vespa.clustercontroller.core.ClusterStateBundle;
+
+public interface ClusterStateBundleCodec {
```
Comment (geirst) [inline clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/ClusterStateBundleCodec.java]: Please consider interface comment.

File: clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/EncodedClusterStateBundle.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core.rpc;
+
+import com.yahoo.compress.Compressor;
+
+public class EncodedClusterStateBundle {
```
Comment (geirst) [inline clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/EncodedClusterStateBundle.java]: Please consider class comment.

File: clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/SlimeClusterStateBundleCodec.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import com.yahoo.compress.Compressor;
+import com.yahoo.slime.*;
+import com.yahoo.vdslib.state.ClusterState;
+import com.yahoo.vespa.clustercontroller.core.AnnotatedClusterState;
+import com.yahoo.vespa.clustercontroller.core.ClusterStateBundle;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class SlimeClusterStateBundleCodec implements ClusterStateBundleCodec {
```
Comment (geirst) [inline clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/rpc/SlimeClusterStateBundleCodec.java]: Please consider class comment.

File: clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/ClusterStateBundleUtil.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core;
+
+import com.yahoo.vdslib.state.ClusterState;
+
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ClusterStateBundleUtil {
```
Comment (geirst) [inline clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/ClusterStateBundleUtil.java]: Please consider class comment.

File: clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/StateMapping.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core;
+
+import com.yahoo.vdslib.state.ClusterState;
+
+public class StateMapping {
```
Comment (geirst) [inline clustercontroller-core/src/test/java/com/yahoo/vespa/clustercontroller/core/StateMapping.java]: Please consider class comment.

**Review by geirst — APPROVED** [review]
👍 Consider adding class comments before merging.

**Comment by vekterli** [comment]
@geirst PTAL

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**